File: src/watcher.rs
use std::{
    path::{Path, PathBuf},
    thread,
};

use anyhow::{Context, Result};
use notify::{
    event::ModifyKind, Config, Event, EventKind, RecommendedWatcher, RecursiveMode, Watcher,
};
use tokio::sync::broadcast::Sender;
use url::Url;

use crate::{
    asset::Asset,
    constants::Paths,
    context::Context as AppContext,
    context_builder::{collect_pages, collect_posts},
    render::{write_asset, write_pages_iter},
    Mode,
};

pub fn start_live_reload(paths: &Paths, context: AppContext, tx: &Sender<crate::Event>) {
    thread::scope(|scope| {
        let css = scope.spawn(|| {
            file_watcher(&paths.styles.canonicalize()?, &["scss"], |event| {
                css_watch_handler(paths, &event, tx)
            })
        });

        let content = scope.spawn(|| {
            file_watcher(&paths.content.canonicalize()?, &["dj", "toml"], |event| {
                content_watch_handler(paths, &event, &context, tx)
            })
        });

        let templates = scope.spawn(|| {
            file_watcher(&paths.templates.canonicalize()?, &["jinja"], |event| {
                content_watch_handler(paths, &event, &context, tx)
            })
        });

        css.join().unwrap().unwrap();
        content.join().unwrap().unwrap();
        templates.join().unwrap().unwrap();
    });
}

fn css_watch_handler(paths: &Paths, event: &Event, tx: &Sender<crate::Event>) -> Result<()> {
    tracing::info!(
        "File(s) {:?} changed, rebuilding CSS",
        strip_prefix_paths(&paths.source, &event.paths)?
    );
    let css = Asset::build_css(paths, Mode::Dev)?;
    write_asset(&paths.out, &css)?;
    tx.send(crate::Event::Reload)?;

    Ok(())
}

fn content_watch_handler(
    paths: &Paths,
    event: &Event,
    context: &AppContext,
    tx: &Sender<crate::Event>,
) -> Result<()> {
    tracing::info!(
        "File(s) {:?} changed, rebuilding site",
        strip_prefix_paths(&paths.source, &event.paths)?
    );
    let url = Url::parse("http://localhost:3000")?;
    let mut pages = collect_pages(paths)?;
    pages.append(&mut collect_posts(paths)?);
    write_pages_iter(
        &paths.out,
        "styles.css",
        Mode::Dev,
        &url,
        &context.templates,
        pages.iter(),
    )?;
    tx.send(crate::Event::Reload)?;

    Ok(())
}

fn strip_prefix_paths(prefix: impl AsRef<Path>, paths: &[PathBuf]) -> Result<Vec<&Path>> {
    paths
        .iter()
        .map(|p| {
            p.strip_prefix(prefix.as_ref().canonicalize()?)
                .context("could not strip prefix")
        })
        .collect()
}

fn file_watcher<F, const N: usize>(path: &Path, extensions: &[&str; N], handler: F) -> Result<()>
where
    F: Fn(Event) -> Result<()>,
{
    let (tx, rx) = std::sync::mpsc::channel();
    let mut watcher = RecommendedWatcher::new(tx, Config::default())?;
    watcher.watch(path, RecursiveMode::Recursive)?;

    for res in rx {
        if let Some(res) = filter_event(res, extensions) {
            handler(res)?;
        }
    }

    Ok(())
}

fn filter_event(res: notify::Result<Event>, extensions: &[&str]) -> Option<Event> {
    match res {
        Ok(event) => match event.kind {
            EventKind::Create(_)
            | EventKind::Modify(ModifyKind::Data(_) | ModifyKind::Name(_))
            | EventKind::Remove(_) => event_has_extension(event, extensions),
            _ => None,
        },
        Err(e) => {
            tracing::error!("watch error: {e:?}");
            None
        }
    }
}

fn event_has_extension(event: Event, extensions: &[&str]) -> Option<Event> {
    if event
        .paths
        .iter()
        .any(|p| path_has_extension(p, extensions))
    {
        Some(event)
    } else {
        None
    }
}

fn path_has_extension(path: &Path, extensions: &[&str]) -> bool {
    path.extension()
        .map_or(false, |e| extensions.contains(&e.to_str().unwrap()))
}

File: src/watcher.ts
import EventEmitter from "node:events";
import { parse } from "node:path";
import { subscribe } from "@parcel/watcher";
import debounce from "debounce";
import { JavaScriptAsset } from "./asset.js";
import { JS_FILES, PATHS } from "./constants.js";
import { Content } from "./content.js";
import { logConfig } from "./logger.js";
import { Site } from "./site.js";

const logger = logConfig.getLogger("watcher");

export class FsEmitter extends EventEmitter {}

const debounceHandler = debounce(async (fn: () => Promise<void>, tx: FsEmitter) => {
	await fn();
	tx.emit("update");
}, 200);

export const startWatcher = (site: Site, tx: FsEmitter): void => {
	void watchTemplates(site, tx);
	void watchContent(site, tx);
	void watchSCSS(site, tx);
	void watchJS(site, tx);
	void watchPublic(site, tx);
};

const watchSCSS = async (site: Site, tx: FsEmitter): Promise<void> => {
	void subscribe(PATHS.styles, (_, events) => {
		console.log(`scss changed: ${events}`);
		for (const _event of events) {
			debounceHandler(async () => {
				logger.info("Rebuilding CSS");
				for await (const asset of site.collectCSS()) {
					await asset.write(site);
				}
			}, tx);
		}
	});
};

const watchJS = async (site: Site, tx: FsEmitter): Promise<void> => {
	void subscribe(PATHS.js, (_, events) => {
		for (const event of events) {
			debounceHandler(async () => {
				logger.info(`Rebuilding JS ${parse(event.path).base}`);
				const item = JS_FILES[event.path as keyof typeof JS_FILES];
				const asset = new JavaScriptAsset(item.source, item.dest);
				site.collectAsset(asset);
				await asset.write(site);
			}, tx);
		}
	});
};

const watchTemplates = async (site: Site, tx: FsEmitter): Promise<void> => {
	void subscribe(PATHS.templates, (_, events) => {
		for (const event of events) {
			debounceHandler(async () => {
				logger.info(`Template ${parse(event.path).base} changed, rebuilding`);
				await site.collectContents();
				await site.writeContent();
				await site.writeSitemap();
			}, tx);
		}
	});
};

const watchContent = async (site: Site, tx: FsEmitter): Promise<void> => {
	void subscribe(PATHS.content, (_err, events) => {
		for (const event of events) {
			debounceHandler(async () => {
				logger.info(`Rebuilding page ${parse(event.path).base}`);
				const content = await Content.fromPath(event.path, "page", site.url);
				await content.write(site);
				site.collectContent(content);
				await site.writeSitemap();
			}, tx);
		}
	});
};

const watchPublic = async (site: Site, tx: FsEmitter): Promise<void> => {
	void subscribe(PATHS.public, (_, events) => {
		for (const event of events) {
			debounceHandler(async () => {
				logger.info(`Rebuilding static file ${parse(event.path).base}`);
				await site.collectStaticFiles();
				await site.copyStaticAssets();
			}, tx);
		}
	});
};
